### PR TITLE
ARROW-477: [Java] Add support for second/microsecond/nanosecond timestamps in-memory and in IPC/JSON layer

### PR DESCRIPTION
**Pull request by alphalfalfa**
Changes include:
- add support for TimeStamp data type with second/microsecond/nanosecond time units
- add an additional readLong() method to timestamp readers to support reading raw long values
- add a simple test case for timestamp readers and writers

### CODE REVIEW

**Comment by alphalfalfa** [comment]
I added TimeStampSec/Micro/Nano and left TimeStamp untouched without renaming it to TimeStampMillis. Should I rename it or not?

File: java/vector/src/main/java/org/apache/arrow/vector/types/Types.java [hunk 1]
```diff
@@ -241,6 +250,23 @@ public FieldWriter getNewFieldWriter(ValueVector vector) {
         return new TimeWriterImpl((NullableTimeVector) vector);
       }
     },
+    // time in microsecond from the Unix epoch, 00:00:00.000000 on 1 January 1970, UTC.
```
Comment (wesm) [inline java/vector/src/main/java/org/apache/arrow/vector/types/Types.java]: typo in comment

File: java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java [hunk 2]
```diff
@@ -561,4 +564,76 @@ public void mapWriterMixedCaseFieldNames() {
...
+    Assert.assertEquals(new DateTime(2L, DateTimeZone.UTC).withZoneRetainFields(DateTimeZone.getDefault()), microDateTime);
+    long microLong = microReader.readLong();
+    Assert.assertEquals(2000L, microLong);
+
+    FieldReader nanoReader = rootReader.reader("nano");
+    nanoReader.setPosition(3);
+    DateTime nanoDateTime = nanoReader.readDateTime();
+    Assert.assertEquals(new DateTime(3L, DateTimeZone.UTC).withZoneRetainFields(DateTimeZone.getDefault()), nanoDateTime);
+    long nanoLong = nanoReader.readLong();
+    Assert.assertEquals(3000000L, nanoLong);
```
Comment (wesm) [inline java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java]: How will you access the long values with the Accessor class?

**Review by wesm — COMMENTED** [review]
@julienledem @jacques-n I can't really opine on the Java API side of things, if you could take a look at this one. As soon as we have support for all the units, we'll want to work on integration tests -- I believe there will be more work on the Java side to handle the metadata. We'll also need to come up with a way to represent nanosecond timestamps that can survive the trip through JSON (since current nanosecond timestamps are above 2^53)

**Comment by julienledem** [comment]
I think renaming will be better for consistency.
@jacques-n ?

**Comment by wesm** [comment]
@jacques-n what do you think (renaming `TimeStamp` to `TimeStampMilli`)? @alphalfalfa could you take a look at my review comments?

File: java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java (hunk 2 above)
Comment (alphalfalfa) [inline java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java]: I am not very sure if I understand the question correctly. I thought accessor.get(index) can be used to read long values while accessor.getObject(index) is used to read DateTime objects.

**Comment by wesm** [comment]
Test failure was caused by GitHub flakiness

File: java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java (hunk 2 above)
Comment (wesm) [inline java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java]: Right. I don't see the `getObject` and `get` behavior tested for nanoseconds. Can you use a test value that exhibits nanosecond truncation to JODA DateTime (if that's what happens with `getObject`)?

**Comment by alphalfalfa** [comment]
@wesm I changed the test value to the epoch nanosecond of 2001-02-03T04:05:06.123456789
@jacques-n what do you think of renaming TimeStamp to TimeStampMilli?

**Comment by wesm** [comment]
@alphalfalfa yes, I think renaming to `TimeStampMilli` is a good idea.

**Comment by wesm** [comment]
@julienledem @jacques-n this looks OK to me outside of the naming question (always the names). Could you take a look when you can?

**Comment by alphalfalfa** [comment]
@wesm, I went ahead and renamed it.

**Comment by julienledem** [comment]
This looks good to me.
@jacques-n @StevenMPhillips more comments?

**Comment by julienledem** [comment]
I think this is good to go.

**Review by julienledem — APPROVED** [review]

**Comment by wesm** [comment]
Thanks @alphalfalfa!

**Pull request closed without merging**